local-cluster/tests/local_cluster.rs

```diff
@@ -5012,3 +5027,401 @@ fn test_boot_from_local_state() {
+        // 1. Vote for duplicate block `D`
+        // 2. See `D` is duplicate, remove from fork choice and reset to ancestor `A`, potentially generating a fork off that ancestor
+        // 3. See `D` is duplicate confirmed, but because of the bug fixed by https://github.com/solana-labs/solana/pull/28172
+        // where we disallow resetting to a slot which matches the last vote slot, we still don't build off `D`,
+        // and continue building on `A`.
+        //
+        // The `target_switch_fork_validator_pubkey` fork is necessary in 2. to force the validator stall trying to switch
+        // vote on that other fork and prevent the validator from making a freebie vote from `A` and allowing consensus to continue.
+
+        // It's important we give the `duplicate_fork_validator1_pubkey` very few leader slots so that:
+        // 1. We have ample time to ensure he doesn't have a chance to make a block until after 2 when they see the block is duplicate.
+        // Otherwise, they'll build the block on top of the duplicate block, which will possibly include a vote for the duplicate block.
+        // We want to avoid this because this will make fork choice pick the duplicate block.
+        // 2. Ensure the `duplicate_fork_validator1_pubkey` sees the target switch fork before it can make another vote
+        // on any forks he himself generates from A. Otherwise, he will make a freebie vote on his own fork from `A` and
+        // consensus will continue on that fork.
+
+        // Give the duplicate fork validator plenty of leader slots after the initial delay to prevent
+        // 1. Switch fork from getting locked out for too long
+        // 2. A lot of consecutive slots in which to build up lockout in tower and make new roots
+        // to resolve the partition
+        (duplicate_fork_validator1_pubkey, 500),
+    ];
+
+    let leader_schedule = create_custom_leader_schedule(validator_to_slots.into_iter());
+
+    // 1) Set up the cluster
+    let (duplicate_slot_sender, duplicate_slot_receiver) = unbounded();
+    let (mut cluster, _validator_keypairs) = test_faulty_node(
+        BroadcastStageType::BroadcastDuplicates(BroadcastDuplicatesConfig {
+            partition: ClusterPartition::Pubkey(vec![
+                duplicate_fork_validator1_pubkey,
+                duplicate_fork_validator2_pubkey,
+            ]),
+            duplicate_slot_sender: Some(duplicate_slot_sender),
+        }),
+        node_stakes,
+        Some(validator_keypairs),
+        Some(FixedSchedule {
+            leader_schedule: Arc::new(leader_schedule),
+        }),
+    );
+
+    // 2) Wait for a duplicate slot to land on both validators and for the target switch
+    // fork validator to get another version of the slot. Also ensure all versions of
+    // the block are playable
+    let mut dup_slot;
+    loop {
+        dup_slot = duplicate_slot_receiver
```

Comment:
What's the purpose of this receiver, doesn't the broadcast stage make duplicates for every slot? Additionally do we expect that within 50 slots we'll get a perfect partition with high probability?

Reply:
> What's the purpose of this receiver, doesn't the broadcast stage make duplicates for every slot

It doesn't make duplicates for every slot, only every `DUPLICATE_RATE` slots. 

>  Additionally do we expect that within 50 slots we'll get a perfect partition with high probability?

I don't think we need a perfect partition now because we only need the leader and the first duplicate validator to have a different version. We then copy that duplicate version to the second duplicate validator

Reply:
got it. As long as the first duplicate validator gets a version that's playable right? so that it can vote on it? Looks like the current duplicate broadcast will always give you a last shred that's playable so not an issue.